File: lint-arwaky/archive/taxonomy/architecture_rule_vo.py
"""architecture_rule_vo — Value objects for architecture rules."""

from pydantic import BaseModel, ConfigDict, Field
from .path_collection_vo import PatternList
from .symbol_collection_vo import PrimitiveTypeList
from .message_status_vo import Count
from .error_code_vo import ErrorCode
from .log_suggestion_vo import BooleanVO, DescriptionVO
from .error_value_vo import ErrorMessage
from .layer_content_vo import LayerNameVO
from .file_suffix_vo import SuffixVO, SuffixPolicyVO


class CustomMessageVO(BaseModel):
    """A pattern-to-message mapping for custom violation messages."""

    model_config = ConfigDict(frozen=True)
    pattern: str  # Regex pattern
    message: ErrorMessage


class MandatoryImportRuleVO(BaseModel):
    """Defines mandatory imports for specific file suffixes."""

    model_config = ConfigDict(frozen=True)
    suffix: SuffixVO
    imports: PatternList


class ArchitectureRule(BaseModel):
    """Rule definition for a layer or set of files."""

    model_config = ConfigDict(arbitrary_types_allowed=True)

    name: DescriptionVO | None = None
    description: DescriptionVO | None = None
    rule_type: ErrorCode = Field(default_factory=lambda: ErrorCode(code="internal"))
    scope: LayerNameVO | PatternList | str = Field(
        default_factory=lambda: LayerNameVO(value="global")
    )

    # Rule parameters
    word_count: Count | None = None
    exceptions: PatternList = Field(
        default_factory=PatternList
    )  # Filenames exempt from this rule
    allowed_import: PatternList = Field(default_factory=PatternList)
    forbidden_import: PatternList = Field(default_factory=PatternList)
    mandatory_import: PatternList = Field(default_factory=PatternList)
    mandatory_import_violation_message: ErrorMessage | None = None
    forbidden_import_violation_message: ErrorMessage | None = None
    suffix_policy: SuffixPolicyVO | None = None
    allowed_suffix: PatternList = Field(default_factory=PatternList)
    forbidden_suffix: PatternList = Field(default_factory=PatternList)

    # Integrity parameters (flattened)
    no_primitives: BooleanVO | PrimitiveTypeList | None = None
    mandatory_imports: list[MandatoryImportRuleVO] | None = None
    barrel_completeness: BooleanVO | None = None
    min_lines: Count | None = None
    max_lines: Count | None = None
    word_count_violation_message: ErrorMessage | None = None
    suffix_violation_message: ErrorMessage | None = None
    no_primitives_violation_message: ErrorMessage | None = None
    min_lines_violation_message: ErrorMessage | None = None
    max_lines_violation_message: ErrorMessage | None = None
    barrel_completeness_violation_message: ErrorMessage | None = None
    forbidden_bypass: PatternList = Field(default_factory=PatternList)
    forbidden_bypass_violation_message: ErrorMessage | None = None
    forbidden_bypass_custom_messages: list[CustomMessageVO] | None = None
    forbid_internal_all: BooleanVO | None = None
    forbid_internal_all_violation_message: ErrorMessage | None = None
    mandatory_class_definition: BooleanVO | None = None
    mandatory_class_definition_violation_message: ErrorMessage | None = None
    dead_inheritance_bypass: BooleanVO | None = None
    dead_inheritance_bypass_violation_message: ErrorMessage | None = None
    dead_inheritance_bypass_custom_messages: list[CustomMessageVO] | None = None
    check_orphan: BooleanVO | None = None
    orphan_entry_points: PatternList = Field(default_factory=PatternList)
    orphan_violation_message: ErrorMessage | None = None
    check_unused_mandatory_imports: BooleanVO | None = None
    check_unused_mandatory_imports_violation_message: ErrorMessage | None = None
    forbidden_inheritance: PatternList = Field(default_factory=PatternList)
    forbidden_inheritance_violation_message: ErrorMessage | None = None

    # Role Integrity Parameters
    no_domain_logic: BooleanVO | None = None
    no_domain_logic_violation_message: ErrorMessage | None = None
    must_implement_service_container_aggregate: BooleanVO | None = None
    must_implement_service_container_aggregate_violation_message: ErrorMessage | None = None
    lazy_eager_initialization_only: BooleanVO | None = None
    lazy_eager_initialization_only_violation_message: ErrorMessage | None = None
    stateless_execution: BooleanVO | None = None
    stateless_execution_violation_message: ErrorMessage | None = None
    single_execution_goal: BooleanVO | None = None
    single_execution_goal_violation_message: ErrorMessage | None = None
    high_level_policy_only: BooleanVO | None = None
    high_level_policy_only_violation_message: ErrorMessage | None = None
    coordinates_multiple_orchestrators: BooleanVO | None = None
    coordinates_multiple_orchestrators_violation_message: ErrorMessage | None = None
    crud_only: BooleanVO | None = None
    crud_only_violation_message: ErrorMessage | None = None
    no_decision_logic: BooleanVO | None = None
    no_decision_logic_violation_message: ErrorMessage | None = None
    thread_async_safe: BooleanVO | None = None
    thread_async_safe_violation_message: ErrorMessage | None = None
    no_domain_data_storage: BooleanVO | None = None
    no_domain_data_storage_violation_message: ErrorMessage | None = None
    owns_system_health_transitions: BooleanVO | None = None
    owns_system_health_transitions_violation_message: ErrorMessage | None = None
    lifecycle_tracking_only: BooleanVO | None = None
    lifecycle_tracking_only_violation_message: ErrorMessage | None = None
    forbid_any_type: BooleanVO | None = None
    forbid_any_type_violation_message: ErrorMessage | None = None


class LegacyLayerRule(BaseModel):
    """Represents a legacy architecture rule (from -> to description)."""

    model_config = ConfigDict(frozen=True)
    source_layer: LayerNameVO
    forbidden_target: LayerNameVO
    description: ErrorMessage = Field(default_factory=lambda: ErrorMessage(value=""))


class LegacyLayerRuleList(BaseModel):
    """Collection of legacy layer rules."""

    model_config = ConfigDict(frozen=True)
    values: list[LegacyLayerRule] = Field(default_factory=list)

    def __iter__(self):
        return iter(self.values)

    def __len__(self) -> int:
        return len(self.values)

File: lint-arwaky/src-rust/taxonomy/architecture_rule_vo.rs
use serde::{Serialize, Deserialize};
use std::collections::{HashMap, HashSet};
use super::*;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ArchitectureRule {
    pub name: DescriptionVO,
    pub description: DescriptionVO,
    pub rule_type: ErrorCode,
    pub scope: LayerNameVO,
    pub word_count: Count,
    pub exceptions: PatternList,
    pub allowed_import: PatternList,
    pub forbidden_import: PatternList,
    pub mandatory_import: PatternList,
    pub mandatory_import_violation_message: ErrorMessage,
    pub forbidden_import_violation_message: ErrorMessage,
    pub suffix_policy: SuffixPolicyVO,
    pub allowed_suffix: PatternList,
    pub forbidden_suffix: PatternList,
    pub no_primitives: BooleanVO,
    pub mandatory_imports: Vec<MandatoryImportRuleVO>,
    pub barrel_completeness: BooleanVO,
    pub min_lines: Count,
    pub max_lines: Count,
    pub word_count_violation_message: ErrorMessage,
    pub suffix_violation_message: ErrorMessage,
    pub no_primitives_violation_message: ErrorMessage,
    pub min_lines_violation_message: ErrorMessage,
    pub max_lines_violation_message: ErrorMessage,
    pub barrel_completeness_violation_message: ErrorMessage,
    pub forbidden_bypass: PatternList,
    pub forbidden_bypass_violation_message: ErrorMessage,
    pub forbidden_bypass_custom_messages: Vec<CustomMessageVO>,
    pub forbid_internal_all: BooleanVO,
    pub forbid_internal_all_violation_message: ErrorMessage,
    pub mandatory_class_definition: BooleanVO,
    pub mandatory_class_definition_violation_message: ErrorMessage,
    pub dead_inheritance_bypass: BooleanVO,
    pub dead_inheritance_bypass_violation_message: ErrorMessage,
    pub dead_inheritance_bypass_custom_messages: Vec<CustomMessageVO>,
    pub check_orphan: BooleanVO,
    pub orphan_entry_points: PatternList,
    pub orphan_violation_message: ErrorMessage,
    pub check_unused_mandatory_imports: BooleanVO,
    pub check_unused_mandatory_imports_violation_message: ErrorMessage,
    pub forbidden_inheritance: PatternList,
    pub forbidden_inheritance_violation_message: ErrorMessage,
    pub no_domain_logic: BooleanVO,
    pub no_domain_logic_violation_message: ErrorMessage,
    pub must_implement_service_container_aggregate: BooleanVO,
    pub must_implement_service_container_aggregate_violation_message: ErrorMessage,
    pub lazy_eager_initialization_only: BooleanVO,
    pub lazy_eager_initialization_only_violation_message: ErrorMessage,
    pub stateless_execution: BooleanVO,
    pub stateless_execution_violation_message: ErrorMessage,
    pub single_execution_goal: BooleanVO,
    pub single_execution_goal_violation_message: ErrorMessage,
    pub high_level_policy_only: BooleanVO,
    pub high_level_policy_only_violation_message: ErrorMessage,
    pub coordinates_multiple_orchestrators: BooleanVO,
    pub coordinates_multiple_orchestrators_violation_message: ErrorMessage,
    pub crud_only: BooleanVO,
    pub crud_only_violation_message: ErrorMessage,
    pub no_decision_logic: BooleanVO,
    pub no_decision_logic_violation_message: ErrorMessage,
    pub thread_async_safe: BooleanVO,
    pub thread_async_safe_violation_message: ErrorMessage,
    pub no_domain_data_storage: BooleanVO,
    pub no_domain_data_storage_violation_message: ErrorMessage,
    pub owns_system_health_transitions: BooleanVO,
    pub owns_system_health_transitions_violation_message: ErrorMessage,
    pub lifecycle_tracking_only: BooleanVO,
    pub lifecycle_tracking_only_violation_message: ErrorMessage,
    pub forbid_any_type: BooleanVO,
    pub forbid_any_type_violation_message: ErrorMessage,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct CustomMessageVO {
    pub pattern: String,
    pub message: ErrorMessage,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct LegacyLayerRule {
    pub source_layer: LayerNameVO,
    pub forbidden_target: LayerNameVO,
    pub description: ErrorMessage,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct LegacyLayerRuleList {
    pub values: Vec<LegacyLayerRule>,
}

impl LegacyLayerRuleList {
    pub fn new(value: Vec<LegacyLayerRule>) -> Self {
        Self { values: value }
    }
    pub fn iter(&self) -> std::slice::Iter<'_, LegacyLayerRule> {
        self.values.iter()
    }
    pub fn len(&self) -> usize {
        self.values.len()
    }
    pub fn is_empty(&self) -> bool {
        self.values.is_empty()
    }
    pub fn push(&mut self, item: LegacyLayerRule) {
        self.values.push(item);
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct MandatoryImportRuleVO {
    pub suffix: SuffixVO,
    pub imports: PatternList,
}

impl ArchitectureRule {
    pub fn new(name: DescriptionVO, description: DescriptionVO, rule_type: ErrorCode, scope: LayerNameVO, word_count: Count, exceptions: PatternList, allowed_import: PatternList, forbidden_import: PatternList, mandatory_import: PatternList, mandatory_import_violation_message: ErrorMessage, forbidden_import_violation_message: ErrorMessage, suffix_policy: SuffixPolicyVO, allowed_suffix: PatternList, forbidden_suffix: PatternList, no_primitives: BooleanVO, mandatory_imports: Vec<MandatoryImportRuleVO>, barrel_completeness: BooleanVO, min_lines: Count, max_lines: Count, word_count_violation_message: ErrorMessage, suffix_violation_message: ErrorMessage, no_primitives_violation_message: ErrorMessage, min_lines_violation_message: ErrorMessage, max_lines_violation_message: ErrorMessage, barrel_completeness_violation_message: ErrorMessage, forbidden_bypass: PatternList, forbidden_bypass_violation_message: ErrorMessage, forbidden_bypass_custom_messages: Vec<CustomMessageVO>, forbid_internal_all: BooleanVO, forbid_internal_all_violation_message: ErrorMessage, mandatory_class_definition: BooleanVO, mandatory_class_definition_violation_message: ErrorMessage, dead_inheritance_bypass: BooleanVO, dead_inheritance_bypass_violation_message: ErrorMessage, dead_inheritance_bypass_custom_messages: Vec<CustomMessageVO>, check_orphan: BooleanVO, orphan_entry_points: PatternList, orphan_violation_message: ErrorMessage, check_unused_mandatory_imports: BooleanVO, check_unused_mandatory_imports_violation_message: ErrorMessage, forbidden_inheritance: PatternList, forbidden_inheritance_violation_message: ErrorMessage, no_domain_logic: BooleanVO, no_domain_logic_violation_message: ErrorMessage, must_implement_service_container_aggregate: BooleanVO, must_implement_service_container_aggregate_violation_message: ErrorMessage, lazy_eager_initialization_only: BooleanVO, lazy_eager_initialization_only_violation_message: ErrorMessage, stateless_execution: BooleanVO, stateless_execution_violation_message: ErrorMessage, single_execution_goal: BooleanVO, single_execution_goal_violation_message: ErrorMessage, high_level_policy_only: BooleanVO, high_level_policy_only_violation_message: ErrorMessage, coordinates_multiple_orchestrators: BooleanVO, coordinates_multiple_orchestrators_violation_message: ErrorMessage, crud_only: BooleanVO, crud_only_violation_message: ErrorMessage, no_decision_logic: BooleanVO, no_decision_logic_violation_message: ErrorMessage, thread_async_safe: BooleanVO, thread_async_safe_violation_message: ErrorMessage, no_domain_data_storage: BooleanVO, no_domain_data_storage_violation_message: ErrorMessage, owns_system_health_transitions: BooleanVO, owns_system_health_transitions_violation_message: ErrorMessage, lifecycle_tracking_only: BooleanVO, lifecycle_tracking_only_violation_message: ErrorMessage, forbid_any_type: BooleanVO, forbid_any_type_violation_message: ErrorMessage,) -> Self {
        Self { name, description, rule_type, scope, word_count, exceptions, allowed_import, forbidden_import, mandatory_import, mandatory_import_violation_message, forbidden_import_violation_message, suffix_policy, allowed_suffix, forbidden_suffix, no_primitives, mandatory_imports, barrel_completeness, min_lines, max_lines, word_count_violation_message, suffix_violation_message, no_primitives_violation_message, min_lines_violation_message, max_lines_violation_message, barrel_completeness_violation_message, forbidden_bypass, forbidden_bypass_violation_message, forbidden_bypass_custom_messages, forbid_internal_all, forbid_internal_all_violation_message, mandatory_class_definition, mandatory_class_definition_violation_message, dead_inheritance_bypass, dead_inheritance_bypass_violation_message, dead_inheritance_bypass_custom_messages, check_orphan, orphan_entry_points, orphan_violation_message, check_unused_mandatory_imports, check_unused_mandatory_imports_violation_message, forbidden_inheritance, forbidden_inheritance_violation_message, no_domain_logic, no_domain_logic_violation_message, must_implement_service_container_aggregate, must_implement_service_container_aggregate_violation_message, lazy_eager_initialization_only, lazy_eager_initialization_only_violation_message, stateless_execution, stateless_execution_violation_message, single_execution_goal, single_execution_goal_violation_message, high_level_policy_only, high_level_policy_only_violation_message, coordinates_multiple_orchestrators, coordinates_multiple_orchestrators_violation_message, crud_only, crud_only_violation_message, no_decision_logic, no_decision_logic_violation_message, thread_async_safe, thread_async_safe_violation_message, no_domain_data_storage, no_domain_data_storage_violation_message, owns_system_health_transitions, owns_system_health_transitions_violation_message, lifecycle_tracking_only, lifecycle_tracking_only_violation_message, forbid_any_type, forbid_any_type_violation_message }
    }
}

impl CustomMessageVO {
    pub fn new(pattern: String, message: ErrorMessage,) -> Self {
        Self { pattern, message }
    }
}

impl MandatoryImportRuleVO {
    pub fn new(suffix: SuffixVO, imports: PatternList,) -> Self {
        Self { suffix, imports }
    }
}

impl LegacyLayerRule {
    pub fn new(source_layer: LayerNameVO, forbidden_target: LayerNameVO, description: ErrorMessage,) -> Self {
        Self { source_layer, forbidden_target, description }
    }
}
